### ISSUE TITLE
Qt 6.5+: docks do not open after the last dock on a panel is closed

### ISSUE THREAD

**Issue by orivej** [musescore/MuseScore#24866]
### Issue type

UI bug (incorrect info or interface appearance)

### Description with steps to reproduce

This issue only happens with Qt 6.5+ (tested on Qt 6.7.2). Previously discussed a bit in the comments of #24097

**Description:** When you open a score and attempt to open a dock that does not have a panel (e.g. you try to open piano or mixer, or you close all of the tabs on the left - palettes, properties, etc. - and try to open one of them back) visually nothing changes, the dock does not open. (@adam248 has also reported this issue as https://github.com/NixOS/nixpkgs/issues/343456)

**Explanation:** KDDockWidgets 1 (including the latest 1.7) are designed to restore docks incompatibly with the way Qt 6.5+ works.

When MuseScore calls `setVisible(true)` in `DockBase::open` to show a previously closed dock widget those panel was also closed, the expected call chain to recreate the panel (frame) is this:
```
muse::dock::DockWindowActionsController::toggleOpened
muse::dock::DockWindow::toggleDock
muse::dock::DockBase::open
KDDockWidgets::QWidgetAdapter::show
QQuickItem::setVisible(true)
QQuickItemPrivate::setVisible(true)
QQuickItemPrivate::setEffectiveVisibleRecur(calcEffectiveVisible()/*=true*/)
QQuickItemPrivate::itemChange
KDDockWidgets::QWidgetAdapter::itemChange
KDDockWidgets::DockWidgetQuick::event
KDDockWidgets::DockWidgetBase::onShown
KDDockWidgets::DockWidgetBase::Private::maybeRestoreToPreviousPosition
KDDockWidgets::DockWidgetBase::Private::restoreToPreviousPosition
KDDockWidgets::LayoutWidget::restorePlaceholder
KDDockWidgets::DefaultWidgetFactory::createFrame
```

However, in Qt 6.5 and later the chain ends at `QQuickItemPrivate::setEffectiveVisibleRecur` since after https://github.com/qt/qtdeclarative/commit/d1b9a4cacfb966cf0a37983d8f8044f3aedf5de3 `calcEffectiveVisible()` returns `false` for an item without a parent (and a dock without a panel has none).

KDDockWidgets have redesigned this in https://github.com/KDAB/KDDockWidgets/commit/764385de9f9dd88c2685e42ef65e1daaaca3d679 which went into the 2.0.0 release. Updating to the latest KDDockWidgets (it still supports Qt 6.2+) seems to be the best way forward.

**Note:** If I revert https://github.com/qt/qtdeclarative/commit/d1b9a4cacfb966cf0a37983d8f8044f3aedf5de3, the closed docks are opened back again, and the only remaining issues with MuseScore 4.4.2 on Qt 6.7.2 that I see (after #24326 is merged) are:

1. The mixer dock is almost blank (only channel strip separators are visible) until I switch to a different dock in the same panel and back (e.g. by pressing 'p' twice to open and close piano keyboard).

2. The navigator dock is a white rectangle the first time it is opened (as long as it was closed when the MuseScore has opened the score) until it is redrawn, e.g. until I pan the view of the score or resize the window. 

### Supporting files, videos and screenshots

![image](https://github.com/user-attachments/assets/fb8e6602-4498-431c-90d0-e7bb2d035b47)

In this state P, F7, F8, F9, F10, F12 do nothing with the UI, while the console prints e.g.
```
| INFO | main_thread | ActionsDispatcher::dispatch | try call action: toggle-piano-keyboard
| INFO | main_thread | ActionsDispatcher::dispatch | try call action: dock-toggle
```

### What is the latest version of MuseScore Studio where this issue is present?

4.4.2

### Regression

No.

### Operating system

NixOS

### Checklist

- [X] This report follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting bugs and issues
- [X] I have verified that this issue has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar issues
- [X] I have attached all requested files and information to this report
- [X] I have attempted to identify the root problem as concisely as possible, and have used minimal reproducible examples where possible

**Comment by orivej** [comment]
I have extracted the custom patches made in MuseScore for KDDockWidgets here https://github.com/orivej/KDDockWidgets/pull/1. This will be useful to make sure that none of the fixes and improvements are lost during update to KDDockWidgets 2

**Comment by cbjeukendrup** [comment]
That's very useful, thanks!
Before we do duplicate work, were you also planning to try updating MuseScore to KDDW2.1, or shall we do that?

**Comment by orivej** [comment]
I'll give it a try - debugging this issue made me learn quite a lot about KDDW and the way MuseScore uses them.

**Comment by cbjeukendrup** [comment]
@orivej That's awesome! I'll assign you for now, but please don't feel pressured in any way.

**Comment by cbjeukendrup** [comment]
@orivej Do you by chance already have a status update about your work on this issue?

**Comment by adam248** [comment]
@cbjeukendrup perhaps this issue needs reassigning to someone else due to a lack of updates.
It has been a month since we have heard back from @orivej which is fine as life does happen to us all.

**Comment by mattfbacon** [comment]
So, is there a workaround? Do I need to downgrade Qt to 6.4?

**Comment by cbjeukendrup** [comment]
We ourselves currently use Qt 6.2.x. I can't really comment on static Qt builds, I have no experience with that. The real solution will be either patching KDDockWidgets v1 (not 100% sure if anyone ever gave this a serious try), or update to v2, as started in https://github.com/musescore/MuseScore/pull/25468 (but we could not finish that for 4.5 yet).

**Comment by darix** [comment]
currently testing 4.5 on openSUSE Tumbleweed. is the "tooltips render broken" (as in way too small to show the text) also a newer Qt problem or does it happen with 6.2.x as well?

**Comment by cbjeukendrup** [comment]
Tooltips issue is introduced by Qt 6.8 as well, but I have a fix already somewhere in #25016. 

Playback toolbar issue has been reported in https://github.com/musescore/MuseScore/issues/21036. It is unclear under what circumstances this issue occurs exactly; in the official AppImage, which uses Qt 6.2.4, the problem doesn't occur, and with Qt 6.8 on macOS it doesn't occur either, so it seems not caused by Qt 6.8 per se.

**Comment by darix** [comment]
hehe I would be happy to apply some testing patches to our package so we can at least get some little bit better working version for openSUSE Tumbleweed. but your patch and the kxdockwidget patch are both so big! :)

**Comment by cbjeukendrup** [comment]
Yes... that's also part of the reason that those patches haven't been integrated yet. 

Regarding KDDockWidgets, i.e. the original issue of this thread, a potential way forward would be to figure out a patch for KDDockWidgets v1, instead of switching to v2 as https://github.com/musescore/MuseScore/pull/25468 does. I believe nobody really investigated this possibility very deeply yet, but something in me says that that may be easier than upgrading to v2. If anyone would like to look into this, see also the KDDockWidgets commit mentioned in the original post here.

**Comment by YellowOnion** [comment]
Please update to 6.2.12 in the mean time, it has heaps of security fixes, you can get it here https://code.qt.io/cgit/qt/qtbase.git/log/?h=6.2

**Comment by cbjeukendrup** [comment]
Of course we already tried updating to 6.2.9/10/11/12, but unfortunately, it introduces unacceptable bugs (randomly disappearing UI elements on Mac, at graphics rendering level, so outside our control) and crashes (on Linux). Therefore we only updated on Windows (6.2.11, because that was the latest available at the time that we did that).

**Comment by Jojo-Schmitz** [comment]
> What security vulnerabilities did the older Musescore versions have?

See https://github.com/musescore/MuseScore/pull/25468#issuecomment-2799917680

**Comment by YellowOnion** [comment]
@mattfbacon all current versions of Musescore ship with outdated and vulnerable versions of Qt, .11 is just safer than .4, but are still security issues, as only the latest version is avaliable to comercial license holders, support for 6.2 publically officially ended in September,  The only secure "versions" are the Linux distro variants that ship with current up to date Qt versions, but those are all broken visually.

![](https://repology.org/badge/vertical-allrepos/qt.svg?exclude_unsupported=1&columns=6)

https://wiki.qt.io/List_of_known_vulnerabilities_in_Qt_products

https://doc.qt.io/qt-6/qt-releases.html#supported-versions

**Comment by mattfbacon** [comment]
Okay, but you are asking people to use a hypothtically secure but broken and unusable version instead of a working but hypothetically vulnerable one. So not surprising that most peope choose the latter. (Note that software that does nothing has no vulnerabilities)

**Comment by darix** [comment]
they assumption that if a distro ships an old version, it is vulnerable. distros do backports. so the screenshot is misleading.

**Comment by darix** [comment]
omg! time to celebrate! cant wait to do this package update

**Comment by cbjeukendrup** [comment]
The docking framework may still be a bit buggy, because I applied some 'dirty' patches to make the current docking code work, rather than updating to KDDW v2 (which we will still do, but later). But at least the basics seem to be working.

**Comment by darix** [comment]
I will celebrate the successes along the way anyway :)

Thank you and the team for all the hard work
